src/other/source_map.rs

```diff
@@ -0,0 +1,159 @@
+use crate::headers::{HeaderName, HeaderValue, Headers, SOURCE_MAP};
+use crate::{bail_status as bail, Status, Url};
+
+use std::convert::TryInto;
+
+/// Links to a file that maps transformed source to the original source.
+///
+/// [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/SourceMap)
+///
+/// # Specifications
+///
+/// - [Source Map Revision 3](https://sourcemaps.info/spec.html)
+///
+/// # Examples
+///
+/// ```
+/// # fn main() -> http_types::Result<()> {
+/// #
+/// use http_types::{Response, Url};
+/// use http_types::other::SourceMap;
+///
+/// let source_map = SourceMap::new(Url::parse("https://example.net/")?);
+///
+/// let mut res = Response::new(200);
+/// source_map.apply(&mut res);
+///
+/// let base_url = Url::parse("https://example.net/")?;
+/// let source_map = SourceMap::from_headers(base_url, res)?.unwrap();
+/// assert_eq!(source_map.location(), &Url::parse("https://example.net/")?);
+/// #
+/// # Ok(()) }
+/// ```
+#[derive(Debug)]
+pub struct SourceMap {
+    location: Url,
+}
+
+impl SourceMap {
+    /// Create a new instance of `SourceMap` header.
+    pub fn new(location: Url) -> Self {
+        Self { location }
+    }
+
+    /// Create a new instance from headers.
+    pub fn from_headers<U>(base_url: U, headers: impl AsRef<Headers>) -> crate::Result<Option<Self>>
+    where
+        U: TryInto<Url>,
+        U::Error: std::fmt::Debug,
+    {
+        let headers = match headers.as_ref().get(SOURCE_MAP) {
+            Some(headers) => headers,
+            None => return Ok(None),
+        };
+
+        // If we successfully parsed the header then there's always at least one
+        // entry. We want the last entry.
+        let header_value = headers.iter().last().unwrap();
+
+        let url = match Url::parse(header_value.as_str()) {
+            Ok(url) => url,
+            Err(_) => match base_url.try_into() {
+                Ok(base_url) => base_url.join(header_value.as_str().trim()).status(400)?,
+                Err(_) => bail!(400, "Invalid base url provided"),
```

Comment:
Also should probably be 500?

Reply:
or probably just not set a status explicitly